kasa/klapprotocol.py

```diff
@@ -0,0 +1,515 @@
+"""Implementation of the TP-Link Smart Home Protocol.
+
+Comment by sdb - 4-Jul-2023
+
+Encryption/Decryption methods based on the works of
+Simon Wilkinson and Chris Weeldon
+
+While working on these changes I discovered my HS100 devices would periodically change their device owner to something that produces the following
+md5 owner hash: 994661e5222b8e5e3e1d90e73a322315.  It seems to be after an update to the on/off state that was scheduled via the app.  Switching the device on and off manually via the
+Kasa app would revert to the correct owner.
+
+For devices that have not been connected to the kasa cloud the theory is that blank username and password md5 hashes will succesfully authenticate but
+at this point I have been unable to verify.
+
+https://gist.github.com/chriswheeldon/3b17d974db3817613c69191c0480fe55
+https://github.com/python-kasa/python-kasa/pull/117
+
+N.B. chrisweeldon implementation had a bug in the encryption logic for determining the initial seq number and Simon Wilkinson's implementation did not seem to support
+incrementing the sequence number for subsequent encryption requests
+
+"""
+import asyncio
+import datetime
+import hashlib
+import logging
+import secrets
+from pprint import pformat as pf
+from typing import Any, Dict, Optional, Tuple, Union
+
+import aiohttp
+from cryptography.hazmat.primitives import hashes, padding
+from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
+from yarl import URL
+
+from .credentials import Credentials
+from .exceptions import AuthenticationException, SmartDeviceException
+from .json import dumps as json_dumps
+from .json import loads as json_loads
+from .protocol import TPLinkProtocol
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class TPLinkKlap(TPLinkProtocol):
+    """Implementation of the KLAP encryption protocol.
+
+    KLAP is the name used in device discovery for TP-Link's new encryption
+    protocol, used by newer firmware versions.
+    """
+
+    DEFAULT_PORT = 80
+    DISCOVERY_PORT = 20002
+    DEFAULT_TIMEOUT = 5
+    DISCOVERY_QUERY = {"system": {"get_sysinfo": None}}
+    TP_SESSION_COOKIE_NAME = "TP_SESSIONID"
+    KASA_SETUP_EMAIL = "kasa@tp-link.net"
+    KASA_SETUP_PASSWORD = "kasaSetup"
+
+    def __init__(
+        self,
+        host: str,
+        credentials: Credentials = Credentials(),
```

Comment:
It might not be that relevant in this specific case, but it's usually better to avoid initializing in function definitions (https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments).

Reply:
Fixed